### PR TITLE
Run Babel asynchronously in `@babel/register`

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Require hooks must run synchronously, but Babel might need to run asynchronously (for example, if there is a `.mjs` config file or plugin).

While it's currently possible to stick to a sync solution (you just need to use a JSON or CJS config file), if we release Babel as ESM it will be impossible to `require("@babel/core")` synchronously before registering the require hook.
We could "solve" it by asking our users to do
```js
import("@babel/register").then(() => require("./entry-point"));
```
instead of
```js
require("@babel/register");
require("./entry-point");
```

but that doesn't solve the problem when using the `node -r` option (https://github.com/babel/babel/issues/11701#issuecomment-780200445).

This PR avoid the problem by moving the compilation to a worker, so that we can load and run Babel asynchronously.

It should be almost 100% backward compatible, except that:
1. It won't be possible to pass function plugins to `@babel/register`'s options, because they aren't serializable. This can be easily solved by moving the plugins to a config file.
2. It requires Node.js ^12.7 (which has `receiveMessageOnPort` support). I set the minimun requirement to ^12.17, which is the version with native ESM support.

I introduced this new implementation at `@babel/register/async` so that we don't introduce breaking changes, but it should become the default in Babel 8.

### CODE REVIEW

File: packages/babel-register/src/async/worker-interface.cjs [hunk 1]
```diff
@@ -0,0 +1,65 @@
+"use strict";
+
+const [major, minor] = process.versions.node.split(".");
+if (+major < 12 || (+major === 12 && +minor < 17)) {
+  throw new Error("@babel/runtime/async requires Node.js >= 12.7");
```
Comment (JLHwung) [inline packages/babel-register/src/async/worker-interface.cjs:5]: ESM is still flagged in [13.0, 13.2).

File: packages/babel-register/test/async.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+      babelrc: true,
+      sourceMaps: false,
+    });
+
+    const result = currentHook(testFileContent, testFile);
+
+    expect(result).toBe('"use strict";\n\nrequire("assert");');
+  });
+
+  test("hook supports .mjs config files", () => {
```
Comment (nicolo-ribaudo) [inline packages/babel-register/test/async.js:197]: Except for this test, this file is a copy of https://github.com/babel/babel/blob/main/packages/babel-register/test/index.js

**Comment by nicolo-ribaudo** [comment]
Ohh C++ errors from Node.js, they are my favourites

**Comment by nicolo-ribaudo** [comment]
Update: this is a bug in Node.js 12.20 (https://github.com/nodejs/node/pull/32623#issuecomment-781668081). It only happens with nested workers (Jest's worker + our worker), and shouldn't affect our users in practice.

I'm pinning to an older v12.x versions for now.

File: packages/babel-register/src/async/worker.cjs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    const result = await babel.transformAsync(code, {
+      ...opts,
+      sourceMaps: opts.sourceMaps === undefined ? "both" : opts.sourceMaps,
+      ast: false,
+    });
+
+    return { type: "result", code: result.code, map: result.map };
+  },
+  __default__(message) {
+    throw new Error("Unkown message: " + message.type);
```
Comment (overlookmotel) [inline packages/babel-register/src/async/worker.cjs:75]: Typo. Should be "Unknown message".

**Pull request closed without merging**